// @nextnode/functions-client@2.0.0/dist/lib/image/utils/quality.js
/**
 * Image quality presets and compression profiles
 * Based on 2025 best practices for optimal file size vs quality balance
 */
/**
 * Default quality settings per format
 * - PNG: 100% (lossless compression)
 * - WebP: 90% (optimal balance for web)
 * - AVIF: 85% (better compression than WebP)
 * - JPEG: 80% (standard web quality)
 */
export const DEFAULT_QUALITY = {
    png: 100,
    webp: 90,
    avif: 85,
    jpeg: 80,
    jpg: 80,
};
/**
 * Lossless compression profile
 * Maximum quality, larger file sizes
 */
export const LOSSLESS_QUALITY = {
    png: 100,
    webp: 100,
    avif: 100,
    jpeg: 95,
    jpg: 95,
};
/**
 * Aggressive compression profile
 * Smaller file sizes, acceptable quality loss
 */
export const AGGRESSIVE_QUALITY = {
    png: 85,
    webp: 75,
    avif: 70,
    jpeg: 65,
    jpg: 65,
};
/**
 * Get quality configuration for a compression profile
 */
export function getQualityProfile(profile) {
    switch (profile) {
        case 'lossless':
            return LOSSLESS_QUALITY;
        case 'aggressive':
            return AGGRESSIVE_QUALITY;
        default:
            return DEFAULT_QUALITY;
    }
}
/**
 * Get quality value for a specific format and profile
 */
export function getQualityForFormat(format, profile = 'balanced', customQuality) {
    if (customQuality !== undefined) {
        return Math.max(0, Math.min(100, customQuality));
    }
    const qualityConfig = getQualityProfile(profile);
    // SVG doesn't have quality setting
    if (format === 'svg') {
        return 100;
    }
    return qualityConfig[format] ?? DEFAULT_QUALITY.webp;
}
/**
 * Determine if format should use mozjpeg optimization
 */
export function shouldUseMozjpeg(format) {
    return format === 'jpeg' || format === 'jpg';
}
/**
 * Determine if format should use pngquant optimization
 */
export function shouldUsePngquant(format) {
    return format === 'png';
}
/**
 * Get Sharp compression options for a format
 */
export function getCompressionOptions(format, quality) {
    switch (format) {
        case 'jpeg':
        case 'jpg':
            return {
                quality,
                mozjpeg: true, // Better compression, slower processing
                progressive: true,
            };
        case 'webp':
            return {
                quality,
                effort: 4, // Balance between speed and compression (0-6)
            };
        case 'avif':
            return {
                quality,
                effort: 4,
            };
        case 'png':
            return {
                quality,
                compressionLevel: 9, // Maximum compression
                adaptiveFiltering: true,
            };
        default:
            return { quality };
    }
}
