// @nextnode/functions-client@2.0.0/dist/lib/image/utils/responsive.js
/**
 * Responsive image utilities for srcset and sizes generation
 * Based on 2025 best practices for responsive images
 */
/**
 * Standard responsive breakpoints (2025 best practices)
 * - 640px: Smartphones
 * - 1024px: Tablets
 * - 1920px: Desktop
 * - 2560px: Large displays (Retina)
 */
export const DEFAULT_BREAKPOINTS = [640, 1024, 1920, 2560];
/**
 * Standard pixel densities for high-DPI displays
 */
export const DEFAULT_DENSITIES = [1, 2, 3];
/**
 * Default responsive image configuration
 */
export const DEFAULT_RESPONSIVE_CONFIG = {
    breakpoints: DEFAULT_BREAKPOINTS,
    densities: DEFAULT_DENSITIES,
    maxWidth: 2560,
};
/**
 * Generate srcset string from array of optimized images
 *
 * @example
 * ```typescript
 * const srcset = generateSrcSet([
 *   { src: '/img-640.webp', width: 640, height: 480, format: 'webp' },
 *   { src: '/img-1024.webp', width: 1024, height: 768, format: 'webp' }
 * ])
 * // Returns: "/img-640.webp 640w, /img-1024.webp 1024w"
 * ```
 */
export function generateSrcSet(images) {
    return images.map(img => `${img.src} ${img.width}w`).join(', ');
}
/**
 * Generate sizes attribute for responsive images
 * Uses mobile-first approach with max-width
 *
 * @example
 * ```typescript
 * generateSizes([
 *   { breakpoint: 640, size: '100vw' },
 *   { breakpoint: 1024, size: '50vw' }
 * ])
 * // Returns: "(max-width: 640px) 100vw, (max-width: 1024px) 50vw, 50vw"
 * ```
 */
export function generateSizes(sizes, defaultSize = '100vw') {
    if (sizes.length === 0) {
        return defaultSize;
    }
    const mediaQueries = sizes.map(({ breakpoint, size }) => `(max-width: ${breakpoint}px) ${size}`);
    return [...mediaQueries, defaultSize].join(', ');
}
/**
 * Calculate responsive widths based on breakpoints and densities
 */
export function calculateResponsiveWidths(breakpoints, densities, maxWidth) {
    const widths = new Set();
    for (const breakpoint of breakpoints) {
        for (const density of densities) {
            const width = breakpoint * density;
            if (!maxWidth || width <= maxWidth) {
                widths.add(width);
            }
        }
    }
    return Array.from(widths).sort((a, b) => a - b);
}
/**
 * Calculate image dimensions maintaining aspect ratio
 */
export function calculateImageDimensions(originalWidth, originalHeight, targetWidth, targetHeight) {
    // If neither dimension specified, return original
    if (!targetWidth && !targetHeight) {
        return { width: originalWidth, height: originalHeight };
    }
    const aspectRatio = originalWidth / originalHeight;
    // If only width specified
    if (targetWidth && !targetHeight) {
        return {
            width: targetWidth,
            height: Math.round(targetWidth / aspectRatio),
        };
    }
    // If only height specified
    if (targetHeight && !targetWidth) {
        return {
            width: Math.round(targetHeight * aspectRatio),
            height: targetHeight,
        };
    }
    // Both specified - use them directly
    return {
        width: targetWidth,
        height: targetHeight,
    };
}
/**
 * Get standard responsive breakpoints with optional custom breakpoints
 */
export function getResponsiveBreakpoints(customBreakpoints) {
    if (customBreakpoints && customBreakpoints.length > 0) {
        return [...customBreakpoints].sort((a, b) => a - b);
    }
    return DEFAULT_BREAKPOINTS;
}
/**
 * Generate common sizes attribute patterns
 */
export const COMMON_SIZES = {
    /** Full width on all screens */
    fullWidth: '100vw',
    /** Half width on desktop, full on mobile */
    halfDesktop: '(max-width: 768px) 100vw, 50vw',
    /** Two columns on desktop, full on mobile */
    twoColumns: '(max-width: 768px) 100vw, 50vw',
    /** Three columns on desktop, two on tablet, full on mobile */
    threeColumns: '(max-width: 640px) 100vw, (max-width: 1024px) 50vw, 33.33vw',
    /** Hero image - full width with max-width constraint */
    hero: '(max-width: 1920px) 100vw, 1920px',
    /** Content image - constrained width */
    content: '(max-width: 768px) 100vw, 768px',
    /** Thumbnail - fixed small size */
    thumbnail: '(max-width: 640px) 150px, 300px',
};
export function buildPictureSources(sources) {
    return sources.map(({ breakpoint, images }) => ({
        media: `(max-width: ${breakpoint}px)`,
        srcSet: generateSrcSet(images),
        type: images[0]?.format ? `image/${images[0].format}` : undefined,
    }));
}
