// @nextnode/config-manager@3.1.0/dist/plugins/vite.js
/**
 * Vite plugin for automatic config type generation
 * Works with Vite, Astro, and other Vite-based build tools
 */
import { existsSync } from 'node:fs';
import { join } from 'node:path';
import { autoGenerateTypes } from '../lib/types/generator.js';
import { typeLogger } from '../lib/utils/logger.js';
export const configManagerPlugin = (options = {}) => {
    const { configDir = 'config', outputFile, watch = true, verbose = false, } = options;
    return {
        name: 'config-manager',
        async buildStart() {
            // Generate types before build
            const generateOptions = {
                configDir,
                force: true, // Always generate on build
                silent: !verbose,
            };
            if (outputFile) {
                generateOptions.outputFile = outputFile;
            }
            const success = await autoGenerateTypes(generateOptions);
            if (!success && verbose) {
                typeLogger.info('No config directory found, skipping type generation', {
                    scope: 'vite-plugin',
                });
            }
        },
        configureServer(server) {
            if (!watch)
                return;
            const configPath = join(server.config.root, configDir);
            if (!existsSync(configPath))
                return;
            // Watch config directory for changes
            server.watcher.add(`${configPath}/*.json`);
            server.watcher.on('change', async (file) => {
                if (file.includes(`${configDir}/`) && file.endsWith('.json')) {
                    if (verbose) {
                        typeLogger.info(`Config file changed: ${file}`, {
                            scope: 'vite-plugin-watch',
                        });
                    }
                    const reloadOptions = {
                        configDir,
                        force: true,
                        silent: !verbose,
                    };
                    if (outputFile) {
                        reloadOptions.outputFile = outputFile;
                    }
                    await autoGenerateTypes(reloadOptions);
                    // Trigger HMR reload
                    server.ws.send({
                        type: 'full-reload',
                    });
                }
            });
            if (verbose) {
                typeLogger.info(`Watching config files in ${configPath}`, {
                    scope: 'vite-plugin-init',
                });
            }
        },
    };
};
// Export for Astro integration
export default configManagerPlugin;
