// @nextnode/logger@0.2.3/dist/core/logger.js
/**
 * Core Logger class for NextNode Logger
 * Zero dependencies, strict TypeScript, production-ready logging with scope support
 */
import { parseLocation, detectEnvironment } from './location.js';
import { generateRequestId } from '../utils/crypto.js';
import { getCurrentTimestamp } from '../utils/time.js';
import { formatLogEntry } from './formatters.js';
// Console methods mapping for type safety
const CONSOLE_METHODS = {
    info: 'log',
    warn: 'warn',
    error: 'error',
};
export class NextNodeLogger {
    environment;
    prefix;
    includeLocation;
    constructor(config = {}) {
        this.environment = config.environment ?? detectEnvironment();
        this.prefix = config.prefix;
        this.includeLocation = config.includeLocation ?? true;
    }
    extractScope(object) {
        if (!object) {
            return { scope: undefined, cleanObject: undefined };
        }
        const { scope, ...rest } = object;
        // Only return cleanObject if there are properties other than scope
        const hasOtherProperties = Object.keys(rest).length > 0;
        return {
            scope: scope ?? undefined,
            cleanObject: hasOtherProperties ? rest : undefined,
        };
    }
    createLogEntry(level, message, object) {
        const { scope, cleanObject } = this.extractScope(object);
        // Add prefix to message if configured
        const finalMessage = this.prefix ? `${this.prefix} ${message}` : message;
        return {
            level,
            message: finalMessage,
            timestamp: getCurrentTimestamp(),
            location: this.includeLocation
                ? parseLocation(this.environment === 'production')
                : { function: 'disabled' },
            requestId: generateRequestId(),
            scope,
            object: cleanObject,
        };
    }
    output(level, formattedMessage) {
        const method = CONSOLE_METHODS[level];
        console[method](formattedMessage);
    }
    log(level, message, object) {
        const entry = this.createLogEntry(level, message, object);
        const formattedMessage = formatLogEntry(entry, this.environment);
        this.output(level, formattedMessage);
    }
    info(message, object) {
        this.log('info', message, object);
    }
    warn(message, object) {
        this.log('warn', message, object);
    }
    error(message, object) {
        this.log('error', message, object);
    }
}
// Factory function for creating logger instances
export const createLogger = (config) => new NextNodeLogger(config);
// Default logger instance for immediate use
export const logger = createLogger();
