// @nextnode/config-manager@3.1.0/dist/lib/config/loader.js
import { readFileSync, existsSync, readdirSync } from 'fs';
import { join } from 'path';
import { deepMerge } from '../utils/helpers.js';
import { validateConfig, getCurrentEnvironment, } from '../utils/validation.js';
import { FILE_EXTENSIONS, CONFIG_CACHE_PREFIX, } from '../definitions/constants.js';
import { DefaultConfigMissingError, InvalidConfigFormatError, InvalidJsonSyntaxError, ConfigDirNotFoundError, } from '../definitions/errors.js';
/**
 * Configuration loader with environment-specific overrides
 */
export class ConfigLoader {
    cache = new Map();
    configDir;
    useCache;
    availableConfigsCache = null;
    constructor(options = {}) {
        // Use provided configDir or default to process.cwd() + '/config'
        this.configDir = options.configDir || join(process.cwd(), 'config');
        // Verify the configuration directory exists
        if (!existsSync(this.configDir)) {
            throw new ConfigDirNotFoundError(this.configDir);
        }
        this.useCache = options.cache ?? true;
    }
    /**
     * Load configuration for the specified environment
     */
    loadConfig(environment) {
        const env = environment || getCurrentEnvironment();
        const cacheKey = this.generateCacheKey(env);
        // Return cached config if available
        if (this.useCache && this.cache.has(cacheKey)) {
            return this.cache.get(cacheKey);
        }
        // Load and merge configurations
        const config = this.mergeConfigs(env);
        // Cache the result
        if (this.useCache) {
            this.cache.set(cacheKey, config);
        }
        return config;
    }
    /**
     * Get the configuration directory path
     */
    getConfigDirectory() {
        return this.configDir;
    }
    /**
     * Clear configuration cache
     */
    clearCache() {
        this.cache.clear();
        this.availableConfigsCache = null;
    }
    /**
     * Check if configuration file exists
     */
    hasConfigFile(filename) {
        const configPath = this.buildConfigPath(filename);
        return existsSync(configPath);
    }
    /**
     * Get available configuration files
     */
    getAvailableConfigs() {
        if (this.availableConfigsCache !== null) {
            return this.availableConfigsCache;
        }
        this.availableConfigsCache = readdirSync(this.configDir)
            .filter(file => file.endsWith(FILE_EXTENSIONS.JSON))
            .map(file => file.replace(FILE_EXTENSIONS.JSON, ''));
        return this.availableConfigsCache;
    }
    /**
     * Merge default and environment configurations
     */
    mergeConfigs(environment) {
        // Load default configuration (required)
        const defaultConfig = this.loadConfigFile('default');
        // Load environment-specific configuration (optional)
        const envConfig = this.loadConfigFile(environment);
        // Merge configurations (environment overrides default)
        return envConfig ? deepMerge(defaultConfig, envConfig) : defaultConfig;
    }
    /**
     * Load a specific configuration file
     */
    loadConfigFile(filename) {
        const configPath = this.buildConfigPath(filename);
        if (!existsSync(configPath)) {
            if (filename === 'default') {
                throw new DefaultConfigMissingError(configPath);
            }
            // Return empty object if environment-specific config doesn't exist
            return {};
        }
        return this.parseConfigFile(configPath);
    }
    /**
     * Parse and validate a configuration file
     */
    parseConfigFile(configPath) {
        const configContent = readFileSync(configPath, 'utf-8');
        let parsedConfig;
        try {
            parsedConfig = JSON.parse(configContent);
        }
        catch (error) {
            if (error instanceof SyntaxError) {
                throw new InvalidJsonSyntaxError(configPath, error.message);
            }
            throw error;
        }
        if (!validateConfig(parsedConfig)) {
            throw new InvalidConfigFormatError(configPath);
        }
        return parsedConfig;
    }
    /**
     * Build full path to configuration file
     */
    buildConfigPath(filename) {
        return join(this.configDir, `${filename}${FILE_EXTENSIONS.JSON}`);
    }
    /**
     * Generate a robust cache key that includes environment and config directory
     * to avoid collisions between different loader instances
     */
    generateCacheKey(environment) {
        const safeDirPath = this.configDir.replace(/[^\w]/g, '_');
        return `${CONFIG_CACHE_PREFIX}${environment}_${safeDirPath}`;
    }
}
